client/mysqltest.cc

```diff
@@ -3837,44 +3837,337 @@ void do_remove_files_wildcard(struct st_command *command)
 }
 
 
+/*
+  SYNOPSIS
+  recursive_copy
+  ds_source      - pointer to dynamic string containing source
+                   directory information
+  ds_destination - pointer to dynamic string containing destination
+                   directory information
+  DESCRIPTION
+  Recursive copy of <ds_source> to <ds_destination>
+
+  Note: When symlink appears, copy the target file
+*/
+
+static int recursive_copy(DYNAMIC_STRING *ds_source,
+                          DYNAMIC_STRING *ds_destination)
+{
+  /* Note that my_dir sorts the list if not given any flags */
+  MY_DIR *src_dir_info= my_dir(ds_source->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+  int error= 0;
+
+  /* Source directory exists */
+  if (src_dir_info)
+  {
+    /* Note that my_dir sorts the list if not given any flags */
+    MY_DIR *dest_dir_info= my_dir(ds_destination->str,
+                                  MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+    /* Create destination directory if it doesn't exist */
+    if (!dest_dir_info)
+    {
+      error= my_mkdir(ds_destination->str, 0777, MYF(0)) != 0;
+      if (error)
+      {
+        my_dirend(dest_dir_info);
+        goto end;
+      }
+    }
+    else
+    {
+      /* Extracting the source directory name */
+      if (ds_source->str[strlen(ds_source->str) - 1] == '/')
+      {
+        ds_source->str[strlen(ds_source->str) - 1]= '\0';
+        ds_source->length= ds_source->length - 1;
+      }
+      char *src_dir_name= strrchr(ds_source->str, '/');
+
+      /* Extracting the destination directory name */
+      if (ds_destination->str[strlen(ds_destination->str) - 1] == '/')
+      {
+        ds_destination->str[strlen(ds_destination->str) - 1]= '\0';
+        ds_destination->length= ds_destination->length - 1;
+      }
+      char *dest_dir_name= strrchr(ds_destination->str, '/');
+
+      /*
+        Destination directory might not exist if source directory
+        name and destination directory name are not same.
+        For example, if source is "abc" and destintion is "def",
+        check for the existance of directory "def/abc". If it exists
+        then, copy the files from source directory(i.e "abc") to
+        destination directory(i.e "def/abc"), otherwise create a new
+        directory "abc" under "def" and copy the files from source to
+        destination directory.
+      */
```

Comment:
uhm. I'd prefer a simpler rule here. That is,
```
  copy_file a/b $dst -r
```
should preferrably either always create `$dst/b` or never create it, even if the destination path is `somewhere/b`